src/main/scala/io/citrine/lolo/transformers/FeatureRotator.scala

```diff
@@ -0,0 +1,225 @@
+                             rotatedFeatures: IndexedSeq[Int],
+                             trans: DenseMatrix[Double]
+                            ) extends Model[PredictionResult[T]] {
+
+  /**
+    * Transform the inputs and then apply the base model
+    *
+    * @param inputs to apply the model to
+    * @return a RotatedFeaturePredictionResult which includes, at least, the expected outputs
+    */
+  override def transform(inputs: Seq[Vector[Any]]): RotatedFeaturePrediction[T] = {
+    val rotatedInputs = FeatureRotator.applyRotation(inputs, rotatedFeatures,  trans)
+    new RotatedFeaturePrediction(baseModel.transform(rotatedInputs), rotatedFeatures, trans)
+  }
+}
+
+/**
+  * Prediction bundling the base learner's prediction with the list of rotated features and the transformation
+  *
+  * @param baseResult
+  * @param rotatedFeatures
+  * @param trans
+  * @tparam T
+  */
+class RotatedFeaturePrediction[T](
+                                  baseResult: PredictionResult[T],
+                                  rotatedFeatures: IndexedSeq[Int],
+                                  trans: DenseMatrix[Double]
+                                 ) extends PredictionResult[T] {
+  /**
+    * Get the expected values for this prediction by delegating to baseResult
+    *
+    * @return expected value of each prediction
+    */
+  override def getExpected(): Seq[T] = baseResult.getExpected().asInstanceOf[Seq[T]]
+
+  /**
+    * Get the uncertainty of the prediction by delegating to baseResult
+    *
+    * @return uncertainty of each prediction
+    */
+  override def getUncertainty(): Option[Seq[Any]] = baseResult.getUncertainty()
+
+  /**
+    * Get the gradient or sensitivity of each prediction
+    *
+    * @return a vector of doubles for each prediction
+    */
+  override def getGradient(): Option[Seq[Vector[Double]]] = {
+    var baseGradient = baseResult.getGradient()
+    // TODO: how should this handle None gradients?
+    baseGradient.map { g =>
+      FeatureRotator.applyRotation(g, rotatedFeatures, trans.t).asInstanceOf[Seq[Vector[Double]]]
+    }
+  }
+
+}
+
+/**
+  * Utilities to compute and apply rotations.
+  */
+object FeatureRotator {
+
+  /**
+    * Draw a random unitary matrix from the uniform (Haar) measure.
+    *
+    * @param dimension for which to get a rotator
+    * @return unitary matrix
+    */
+  def getRandomRotation(dimension: Int): DenseMatrix[Double] = {
+    val X = DenseMatrix.rand(dimension, dimension, Gaussian(0, 1))
+    val QR(_Q, _R) = qr(X)
+    diag(signum(diag(_R))) * _Q.toDenseMatrix
+  }
+
+  /**
+    * Get list of feature indices that make sense to rotate.
+    *
+    * @param rep representative vector of features
+    * @return list of feature indices that are doubles
+    */
+  def getDoubleFeatures(rep: Vector[Any]): IndexedSeq[Int] = {
+    rep.indices.collect( i =>
+      rep(i) match {
+        case x if x.isInstanceOf[Double] => i
```

Comment:
Does `case _: Double => i` not work because of type erasure?

I'd be a bit surprised if this collect syntax works.  Can you test it when there are some categorical inputs?  You can use [this pattern](https://github.com/CitrineInformatics/lolo/blob/develop/src/test/scala/io/citrine/lolo/learners/RandomForestTest.scala#L18) to generate mixed real/categorical input vectors.

Reply:
This was an experiment that I neglected to roll back, trying to fix IntelliJ's complaint that a `PartialFunction[Int,NotInferredB]` is required. I now suspect that complaint reflects a static analysis bug, but I'd appreciate if you could check this spot for potential issues the static analyzer is correctly flagging.

Reply:
`rep.indices.collect` allows you to collect the elements of `rep.indices` that meet some criterion, but `rep.indices` is just a list of indices (0, 1, 2, 3, ...) and even though that's what you're collecting, you're filtering on something else (the type of `rep(i)`).
I think the simplest way to do this is 
```rep.indices.filter(i => rep(i).isInstanceOf[Double])```
If you want to avoid `isInstanceOf`, then you could do
```
rep.zip(rep.indices).collect {
      case (_: Double, i: Int) => i
    }
```

Reply:
I do prefer that. Thanks!